botocore/utils.py

```diff
@@ -611,20 +600,138 @@ def retrieve_iam_role_credentials(self):
             token=token,
         )
-        return json.loads(r.text)
+        return r
+
+    def _resolve_get_credentials(self, role_name, token):
+        try:
+            retry_count = 0
+            while retry_count < 2:
+                try:
+                    credentials_response = self._get_credentials(
+                        role_name, token
+                    )
+                    if credentials_response.status_code == 200:
+                        credentials = json.loads(credentials_response.text)
+                        if self._use_extended_api is None:
+                            self._use_extended_api = True
+                        if self._contains_all_credential_fields(credentials):
+                            return_credentials = {
+                                'role_name': role_name,
+                                'access_key': credentials['AccessKeyId'],
+                                'secret_key': credentials['SecretAccessKey'],
+                                'token': credentials['Token'],
+                                'expiry_time': credentials['Expiration'],
+                            }
+                            if self._use_extended_api is True:
+                                if 'AccountId' in credentials:
+                                    return_credentials['account_id'] = (
+                                        credentials['AccountId']
+                                    )
+                        else:
+                            # IMDS can return a 200 response that has a JSON formatted
+                            # error message (i.e. if ec2 is not trusted entity for the
+                            # attached role). We do not necessarily want to retry for
+                            # these and we also do not necessarily want to raise a key
+                            # error. So at least log the problematic response and return
+                            # an empty dictionary to signal that it was not able to
+                            # retrieve credentials. These error will contain both a
+                            # Code and Message key.
+                            if (
+                                'Code' in credentials
+                                and 'Message' in credentials
+                            ):
+                                logger.debug(
+                                    'Error response received when retrieving'
+                                    'credentials: %s.',
+                                    credentials,
+                                )
+                            return {}
+                        self._evaluate_expiration(return_credentials)
+                        return return_credentials
+                    else:
+                        self._perform_fallback()
+                except self._RETRIES_EXCEEDED_ERROR_CLS:
+                    self._perform_fallback()
```

Comment:
Can you explain why we're calling `self._perform_fallback()` in the case of a retries exceeded error?